crates/autopilot/src/solvable_orders.rs

```diff
@@ -657,22 +657,22 @@ fn to_normalized_price(price: f64) -> Option<U256> {
 }
 
 async fn filter_unsupported_tokens(
-    mut orders: Vec<Order>,
-    bad_token: &dyn BadTokenDetecting,
+    orders: Vec<Order>,
+    bad_token: Arc<dyn BadTokenDetecting>,
 ) -> Result<Vec<Order>> {
-    // Can't use normal `retain` or `filter` because the bad token detection is
-    // async. So either this manual iteration or conversion to stream.
-    let mut index = 0;
-    'outer: while index < orders.len() {
-        for token in orders[index].data.token_pair().unwrap() {
-            if !bad_token.detect(token).await?.is_good() {
-                orders.swap_remove(index);
-                continue 'outer;
+    try_join_all(orders.into_iter().map(|order| {
+        let bad_token = bad_token.clone();
+        async move {
+            for token in order.data.token_pair().unwrap() {
+                if !bad_token.detect(token).await?.is_good() {
+                    return Ok(None);
+                }
```

Comment:
This is slower than it could be for 2 reasons:
1. we go over all orders which will have tonnes of duplicate tokens
2. we go over the 2 tokens of each order serially

Instead I suggest collecting all unique traded tokens into a set before and then run bad token detection for all of them in parallel.

Reply:
This is actually a really good point.

Reply:
Oh, true. How come I missed that? Thanks!